### ISSUE TITLE
Why change slim pruner to iterative pruner?

### ISSUE THREAD

**Issue by twmht** [microsoft/nni#3841]
**Describe the issue**:
The original implementation of slim pruner is a one shot pruner (https://github.com/foolwood/pytorch-slimming). 

They regularize batch normalization to train a model first. and then prune the model based on the final batch normalization statistic. so it's a one shot pruner.

So why you change slim pruner to iterative pruner (https://github.com/microsoft/nni/blob/master/nni/algorithms/compression/pytorch/pruning/iterative_pruner.py#L393)? did you test the accuracy?

**Comment by J-shang** [comment]
Hi @twmht, thanks for your issue, it is a good question. We put slim pruner into iterative pruner because we want to handle the sparsity regularization process (sparsify gamma in batch normalization in slim) in nni. And in the future, we also want to handle the retrain (finetune) process to achieve true iteration as described in the paper (https://arxiv.org/pdf/1708.06519.pdf). 

![image](https://user-images.githubusercontent.com/33053116/122198472-c42e8c80-cecb-11eb-93a5-6785086c6210.png)

We are currently planning how to better unify the iterative process to meet the needs of most pruning algorithms. If you have any ideas or suggestions, welcome to discuss with us.

**Comment by twmht** [comment]
@J-shang 

Thank you for the helpful comment.

So 
 `epochs_per_iteration`  in https://github.com/microsoft/nni/blob/master/nni/algorithms/compression/pytorch/pruning/iterative_pruner.py#L81  means that the number of multi-pass?

But in my opinion, most of one shot pruners can be trained with multi pass scheme, did you train with  L1FilterPruner  with iterative manner?

**Comment by J-shang** [comment]
@twmht 

Yes, `epochs_per_iteration` means the number of multi-pass.

You are right, most one-shot pruners in nni have an iterative scheme (L1Filter, FPGM), we will support the iterative manner for these pruners in the future. The reason why we migrated slim at first is that we think the sparsify process belongs to the pruning algorithm.

**Comment by chumingqian** [comment]
Hi, @twmht :
      I still don't understand,  for the slim pruner,  currently,   they still don't regularize batch normalization to train a model first,   am  I  missing sth?

**Comment by twmht** [comment]
@chumingqian 

they did  regularize batch normalization

**Comment by chumingqian** [comment]
> @chumingqian
> 
> they did regularize batch normalization



> @chumingqian
> 
> they did regularize batch normalization

Thank you for your quick reply,  could you show me the specific code for regularizing the loss item,  and add the regularization loss item with the normal loss item.
 I only find the following :
https://nni.readthedocs.io/zh/stable/_modules/nni/algorithms/compression/pytorch/pruning/iterative_pruner.html#SlimPruner

Thanks in advace.

**Comment by twmht** [comment]
@chumingqian 

you don't need to actually add loss item, this is the equivalent when doing backward propagation.

```
 def _callback(self):
        for _, wrapper in enumerate(self.get_modules_wrapper()):
            wrapper.module.weight.grad.data.add_(self.scale * torch.sign(wrapper.module.weight.data))

```

**Comment by chumingqian** [comment]
> @chumingqian
> 
> you don't need to actually add a loss item, this is the equivalent when doing backward propagation.
> 
> ```
>  def _callback(self):
>         for _, wrapper in enumerate(self.get_modules_wrapper()):
>             wrapper.module.weight.grad.data.add_(self.scale * torch.sign(wrapper.module.weight.data))
> ```

Thank you very much,   can I ask the last question,   why this is equivalent when doing backward propagation?  What is the mathematical theory supported?
